patroni/postgresql/callback_executor.py

```diff
@@ -13,28 +14,59 @@ def __init__(self):
         self._lock = Lock()
         self._cmd = None
         self._process = None
+        self._process_cmd = None
+        self._process_children = []
         self._callback_event = Event()
         self.start()
 
     def call(self, cmd):
+        self._cancel()
+        self._cmd = cmd
+        self._callback_event.set()
+
+    def _cancel(self):
         with self._lock:
-            if self._process and self._process.poll() is None:
+            if self._process and self._process.poll() is None and not self._process_children:
+                try:
+                    self._process_children = self._process.children(recursive=True)
+                except psutil.Error:
+                    pass
+
                 try:
                     self._process.kill()
-                    logger.warning('Killed the old callback process because it was still running: %s', self._cmd)
-                except OSError:
+                    logger.warning('Killed the old callback process because it was still running: %s',
+                                   self._process_cmd)
+                except psutil.NoSuchProcess:
+                    return
+                except psutil.AccessDenied:
                     logger.exception('Failed to kill the old callback')
-        self._cmd = cmd
-        self._callback_event.set()
+
+    def _wait(self):
+        self._process.wait()
+
+        waitlist = []
+        with self._lock:
+            for child in self._process_children:
+                if child.is_running():
+                    try:
+                        child.kill()
```

Comment:
why directly kill and not terminate followed by a kill later ? that can (at least in theory) enable callback children to terminate gracefully

Reply:
We really want to kill them, as fast as possible.